File: backend/GameApp.Domain/ValueObjects/Enemies/CriticalDamage.cs
namespace GameApp.Domain.ValueObjects.Combat;

// Value Object of Critical Damage
public class CriticalDamage
{
    private static readonly string _invalidProbabilityMessage =
        "Critical probability must be between 0 and 100";

    private static readonly string _invalidExtraDamageMessage =
        "Extra critical damage must be greater than or equal to 0";

    private readonly int CriticalProbability;
    private readonly int ExtraDamage;

    // Constructor
    public CriticalDamage(int criticalProbability, int extraDamage)
    {
        if (criticalProbability < 0 || criticalProbability > 100)
            throw new ArgumentException(_invalidProbabilityMessage, nameof(criticalProbability));

        if (extraDamage < 0)
            throw new ArgumentException(_invalidExtraDamageMessage, nameof(extraDamage));

        CriticalProbability = criticalProbability;
        ExtraDamage = extraDamage;
    }

    // Getters
    public int GetCriticalProbability() => CriticalProbability;
    public int GetExtraDamage() => ExtraDamage;

    // setters
    public CriticalDamage SetCriticalProbability(int newProbability)
    {
        return new CriticalDamage(newProbability, ExtraDamage);
    }

    public CriticalDamage SetExtraDamage(int newExtraDamage)
    {
        return new CriticalDamage(CriticalProbability, newExtraDamage);
    }

    // Equality
    public override bool Equals(object? obj)
    {
        if (obj is not CriticalDamage other)
            return false;

        return CriticalProbability == other.CriticalProbability
            && ExtraDamage == other.ExtraDamage;
    }

    public override int GetHashCode()
        => HashCode.Combine(CriticalProbability, ExtraDamage);

    public override string ToString()
        => $"CriticalDamage(probability={CriticalProbability}%, extraDamage={ExtraDamage})";
}

File: frontend/gameapp-frontend/src/domain/value-objects/enemies/critical-damage.ts
// Value Object for critical damage
export class CriticalDamage {
    static ERROR_INVALID_PROBABILITY =
        "Critical probability must be between 0 and 100";

    static ERROR_INVALID_EXTRA_DAMAGE =
        "Extra critical damage must be greater than or equal to 0";

    private readonly _criticalProbability: number;
    private readonly _extraDamage: number;

    constructor(criticalProbability: number, extraDamage: number) {
        if (criticalProbability < 0 || criticalProbability > 100) {
            throw new Error(CriticalDamage.ERROR_INVALID_PROBABILITY);
        }

        if (extraDamage < 0) {
            throw new Error(CriticalDamage.ERROR_INVALID_EXTRA_DAMAGE);
        }

        this._criticalProbability = criticalProbability;
        this._extraDamage = extraDamage;
    }

    // getters
    get criticalProbability(): number {
        return this._criticalProbability;
    }

    get extraDamage(): number {
        return this._extraDamage;
    }

    // immutable setters
    setCriticalProbability(newProbability: number): CriticalDamage {
        return new CriticalDamage(newProbability, this._extraDamage);
    }

    setExtraDamage(newExtraDamage: number): CriticalDamage {
        return new CriticalDamage(this._criticalProbability, newExtraDamage);
    }

    // equality
    equals(other: unknown): boolean {
        return other instanceof CriticalDamage &&
            this._criticalProbability === other._criticalProbability &&
            this._extraDamage === other._extraDamage;
    }

    toString(): string {
        return `CriticalDamage(probability=${this._criticalProbability}%, extraDamage=${this._extraDamage})`;
    }
}
